### PR TITLE
compat with .NET >=2 on Windows

### PR DESCRIPTION
**Pull request by prepare**

### CODE REVIEW

File: NOpenType/Glyph.cs [hunk 1]
```diff
@@ -21,24 +32,31 @@ public Point(short x, short y, bool on)
...
-    
+
     public class Glyph
     {
         private readonly short[] _x;
         private readonly short[] _y;
         private readonly bool[] _on;
         private readonly ushort[] _contourEndPoints;
         private readonly Bounds _bounds;
+        private Flag[] _flags;
```
Comment (vidstige) [inline NOpenType/Glyph.cs]: There is no need to store the flags. Only the "on" property is relevant. The other parts of the flags is only for parsing ttf.

File: NOpenType/Glyph.cs [hunk 2]
```diff
@@ -21,24 +32,31 @@ public Point(short x, short y, bool on)
...
+        public static readonly Glyph Empty = new Glyph(new short[0], new short[0], new Flag[0], new ushort[0], Bounds.Zero);
 
-        public static readonly Glyph Empty = new Glyph(new short[0], new short[0], new bool[0], new ushort[0], Bounds.Zero);
-        
-        public Glyph(short[] x, short[] y, bool[] on, ushort[] contourEndPoints, Bounds bounds)
+        public Glyph(short[] x, short[] y, Flag[] flags, ushort[] contourEndPoints, Bounds bounds)
         {
+            //bool[] on
```
Comment (vidstige) [inline NOpenType/Glyph.cs]: remove comment.

File: NOpenType/Glyph.cs [hunk 3]
```diff
@@ -21,24 +32,31 @@ public Point(short x, short y, bool on)
...
-            _on = on;
+            _flags = flags;
             _contourEndPoints = contourEndPoints;
             _bounds = bounds;
+
+            _on = new bool[flags.Length];
+            for (int i = flags.Length - 1; i >= 0; --i)
+            {
+                _on[i] = ((int)_flags[i] & 0x1) == 1;
+            }
```
Comment (vidstige) [inline NOpenType/Glyph.cs]: Remove _flags.

File: NOpenType/Glyph.cs [hunk 4]
```diff
@@ -133,5 +177,6 @@ private int GetContourEnd(int contourIndex)
         public short[] Y { get { return _y; } }
         public bool[] On { get { return _on; } }
         public ushort[] EndPoints { get { return _contourEndPoints; } }
+        public Flag[] Flags { get { return _flags; } }
```
Comment (vidstige) [inline NOpenType/Glyph.cs]: remove flags.

File: NOpenType/GlyphPathBuilderBase.cs [hunk 5]
```diff
@@ -18,7 +17,7 @@ public GlyphPathBuilderBase(Typeface typeface, IGlyphRasterizer rasterizer)
         }
         const double FT_RESIZE = 64; //essential to be floating point
 
-        void RenderGlyph(ushort[] contours, FtPoint[] ftpoints, int[] flags)
+        void RenderGlyph(ushort[] contours, FtPoint[] ftpoints, Flag[] flags)
```
Comment (vidstige) [inline NOpenType/GlyphPathBuilderBase.cs]: Go for bool and then use `On` below.

File: NOpenType/GlyphPathBuilderBase.cs [hunk 6]
```diff
@@ -215,8 +214,9 @@ void RenderGlyph(Glyph glyph)
             {
                 ftpoints.Add(new FtPoint(xs[i], ys[i]));
             }
-            var flags = glyph.On.Select(on => on ? 0 : 1).ToArray();
-            RenderGlyph(glyph.EndPoints, ftpoints.ToArray(), flags);
+            //var flags = glyph.On.Select(on => on ? 0 : 1).ToArray();
+
+            RenderGlyph(glyph.EndPoints, ftpoints.ToArray(), glyph.Flags);
```
Comment (vidstige) [inline NOpenType/GlyphPathBuilderBase.cs]: Just use `On` instead. Much simpler.

File: NOpenType/Interpreter.cs [hunk 7]
```diff
@@ -70,7 +69,16 @@ private Instruction[] BuildLookup(List<Instruction> instructions)
...
+                //result[opcode] = instructions.First(i => i.Matches((byte)opcode));
+
+                foreach (Instruction inst in instructions)
+                {
+                    if (inst.Matches((byte)opcode))
+                    {
+                        result[opcode] = inst;
+                        break;
+                    }
+                }
```
Comment (vidstige) [inline NOpenType/Interpreter.cs:81]: Use `instructions.Find` instead. This is the drawback without linq :-/

File: NOpenType/Net20/ExtensionAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿#define NET20
+namespace System.Runtime.CompilerServices
+{
+#if NET20
+    public partial class ExtensionAttribute : Attribute { }
+#endif
+}
```
Comment (vidstige) [inline NOpenType/Net20/ExtensionAttribute.cs:7]: 1. add trailing whitespace.
2. Why is this file needed?

File: NOpenType/OpenTypeReader.cs [hunk 9]
```diff
@@ -2,16 +2,22 @@
...
         {
-            return tables.Single(t => t.Tag == tableName);
+            foreach (TableEntry t in tables)
+            {
+                if(t.Tag == tableName)
+                {
+                    return t;
+                }
+            }
+            return null;
```
Comment (vidstige) [inline NOpenType/OpenTypeReader.cs]: Single raises an exception if there are no matchin entries. Instead make a function that searcher in an enumerable.

File: NRasterizer.CLI/NRasterizer.CLI.csproj.user [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|AnyCPU'">
+    <StartArguments>tahoma.ttf test.png</StartArguments>
+  </PropertyGroup>
+</Project>
```
Comment (vidstige) [inline NRasterizer.CLI/NRasterizer.CLI.csproj.user:6]: Don't commit this file.

File: NRasterizer.CLI/Program.cs [hunk 11]
```diff
@@ -27,12 +27,17 @@ private static void BlitTo(Raster raster, Bitmap bitmap)
...
-                    si += raster.Stride;
-                    di += bitmapData.Stride;
+                    IntPtr di = bitmapData.Scan0;
+                    byte* di1 = (byte*)di;
+                    for (int y = 0; y < b.Height; y++)
+                    {
+                        Marshal.Copy(raster.Pixels, si, (IntPtr)di1, b.Width);
+                        si += raster.Stride;
+                        di1 += bitmapData.Stride; 
+                    }
```
Comment (vidstige) [inline NRasterizer.CLI/Program.cs]: Remove unsafe block. Better to use safe code here as it is not performance critical.

**Comment by vidstige** [comment]
Thank you!
1. Some comments to adress
2. Test is failing. Why? Try to make green if possible. I will change the tests so that they instead use the new and shiny interface instead, but for now please make then green.
3. Rebase branch, too many master merges :-) See below

## How to rebase
1. `git checkout -b dotnet-2.0-rebased` # create new branch to avoid force pushing
2. `git fetch upstream` # make sure you've got latest master
3. `git rebase upstream/master`  # rebase, this will remove master merge commits.
4. `git push` # push to github
5. Create new pull request from new branch.

File: NOpenType/OpenTypeReader.cs [hunk 12]
```diff
@@ -2,16 +2,22 @@
...
 namespace NRasterizer
 {
     public class OpenTypeReader
     {
         private TableEntry FindTable(IEnumerable<TableEntry> tables, string tableName)
         {
-            return tables.Single(t => t.Tag == tableName);
+            foreach (TableEntry t in tables)
+            {
+                if(t.Tag == tableName)
```
Comment (vidstige) [inline NOpenType/OpenTypeReader.cs]: also space after `if`

File: NOpenType/Net20/ExtensionAttribute.cs (hunk 8 above)
Comment (prepare) [reply NOpenType/Net20/ExtensionAttribute.cs:7]: on Windows with .NET 2.0,
no ExtensionAttribute class, so => can't use extension method :(
add a 'fake' one here, then a C# compile can use extension method.

File: NOpenType/Interpreter.cs (hunk 7 above)
Comment (prepare) [reply NOpenType/Interpreter.cs:81]: ok

File: NOpenType/Glyph.cs (hunk 1 above)
Comment (prepare) [reply NOpenType/Glyph.cs]: ok

File: NOpenType/Glyph.cs (hunk 3 above)
Comment (prepare) [reply NOpenType/Glyph.cs]: ok

File: NOpenType/GlyphPathBuilderBase.cs (hunk 5 above)
Comment (prepare) [reply NOpenType/GlyphPathBuilderBase.cs]: ok

File: NRasterizer.CLI/Program.cs (hunk 11 above)
Comment (prepare) [reply NRasterizer.CLI/Program.cs]: The fixed the err on .NET20,
err_on_net20

but not found on .net >=4

---

but is ok for test project to run on .net >=4 
so I will remove this.

File: NRasterizer.CLI/NRasterizer.CLI.csproj.user (hunk 10 above)
Comment (prepare) [reply NRasterizer.CLI/NRasterizer.CLI.csproj.user:6]: sorry, haha
I will remove this

File: NOpenType/GlyphPathBuilderBase.cs (hunk 5 above)
Comment (prepare) [reply NOpenType/GlyphPathBuilderBase.cs]: If I use on[], I must create a new array.
but If I use Flags on new array create.
what do you think?

**Comment by vidstige** [comment]
You can (and should )also run test locally from visual studio :+1:

File: NOpenType/GlyphPathBuilderBase.cs (hunk 5 above)
Comment (vidstige) [reply NOpenType/GlyphPathBuilderBase.cs]: Create new array 👍

File: NOpenType/Net20/ExtensionAttribute.cs (hunk 8 above)
Comment (vidstige) [reply NOpenType/Net20/ExtensionAttribute.cs:7]: Cool! :-)

File: NRasterizer.CLI/NRasterizer.CLI.csproj.user (hunk 10 above)
Comment (vidstige) [reply NRasterizer.CLI/NRasterizer.CLI.csproj.user:6]: :-D

File: NRasterizer.CLI/Program.cs (hunk 11 above)
Comment (vidstige) [reply NRasterizer.CLI/Program.cs]: ah, right... :-/ hmm. Yeah, let's go with .net >= 4 for now then.

There are some tricks around it if we really need it.

File: NOpenType/GlyphPathBuilderBase.cs (hunk 6 above)
Comment (prepare) [reply NOpenType/GlyphPathBuilderBase.cs]: err1_on_net20

**Comment by prepare** [comment]
@vidstige 
Test on local machine
I can't fix the error that occurs in xaml project.
I'm not familiar with xaml :(
but I think its because of namespace name
of existing rendering elements have changed to
NRasterizer.NRasterizer

err2_on_net20

do you have any idea it fix it?

File: NOpenType/Glyph.cs [hunk 13]
```diff
@@ -1,9 +1,20 @@
...
+    {
+        ControlPoint = 0,
+        OnCurve = 1,
+        XByte = 2,
+        YByte = 4,
+        Repeat = 8,
+        XSignOrSame = 16,
+        YSignOrSame = 32
+    }
+
```
Comment (vidstige) [inline NOpenType/Glyph.cs:6]: Move back inside class.

File: NOpenType/GlyphPathBuilderBase.cs [hunk 14]
```diff
@@ -103,43 +104,27 @@ void RenderGlyph(ushort[] contours, FtPoint[] ftpoints, int[] flags)
...
-                                        thirdControlPoint = new FtPointD(vpoint);
-                                    }
-                                    else
-                                    {
-                                        //printf("[%d] bzc2nd,  x: %d,y:%d \n", mm, vpoint.x, vpoint.y);
-                                        secondControlPoint = new FtPointD(vpoint);
-                                    }
+                                {
+                                    secondControlPoint = new FtPointD(vpoint_x, vpoint_y);
                                 }
```
Comment (vidstige) [inline NOpenType/GlyphPathBuilderBase.cs:110]: What happened here?? Please do one thing per pull request, otherwise this will take 100 years...

File: NOpenType/GlyphPathBuilderBase.cs [hunk 15]
```diff
@@ -103,43 +104,27 @@ void RenderGlyph(ushort[] contours, FtPoint[] ftpoints, int[] flags)
...
+                                    //----------
+                                    //generate curve3
+                                    _rasterizer.Curve3(secondControlPoint.x / FT_RESIZE, secondControlPoint.y / FT_RESIZE,
+                                        mid.x / FT_RESIZE, mid.y / FT_RESIZE);
+                                    //------------------------
+                                    controlPointCount--;
+                                    //------------------------
+                                    //printf("[%d] bzc2nd,  x: %d,y:%d \n", mm, vpoint.x, vpoint.y);
+                                    secondControlPoint = new FtPointD(vpoint_x, vpoint_y);
+
```
Comment (vidstige) [inline NOpenType/GlyphPathBuilderBase.cs:128]: Algorithm fixes. For next time, please make in separate pull request...

File: NOpenType/Tables/Glyf.cs [hunk 16]
```diff
@@ -95,15 +85,21 @@ private static Glyph ReadSimpleGlyph(BinaryReader input, int count, Bounds bound
             var xs = ReadCoordinates(input, pointCount, flags, Flag.XByte, Flag.XSignOrSame);
             var ys = ReadCoordinates(input, pointCount, flags, Flag.YByte, Flag.YSignOrSame);
 
-            return new Glyph(xs, ys, flags.Select(f => HasFlag(f, Flag.OnCurve)).ToArray(), endPoints, bounds);
+            var onCurves = new bool[flags.Length];
+            for (int i = flags.Length - 1; i >= 0; --i)
+            {
+                onCurves[i] = ((int)flags[i] & 0x1) == 1;
```
Comment (vidstige) [inline NOpenType/Tables/Glyf.cs]: 1. Don't cast to int
2. Bitwise `and` with the flag constant instead like so

`flags[i] & Flags.OnCurve == Flags.OnCurve`

File: NRasterizer.CLI/NRasterizer.CLI.csproj [hunk 17]
```diff
@@ -18,6 +19,7 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
     <Externalconsole>true</Externalconsole>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (vidstige) [inline NRasterizer.CLI/NRasterizer.CLI.csproj]: Remove this

**Comment by vidstige** [comment]
The WPF is not being built by build server. I introduced those errors.... Click "Details" to see actual error message. One or more unit tests fail... 

See the following link on how to run tests locally: http://nunit.org/index.php?p=vsTestAdapter&r=2.6.4

**Comment by vidstige** [comment]
You can also run from command line like on travis. See .travis.yml file for more info. But I recommend installing the NUnit test adapter.

**Comment by vidstige** [comment]
I merged my branch. I added x and y parameters to decide where to draw letter. Needed for multi-char support. Should be an easy merge. Sorry for inconvenience.

**Comment by prepare** [comment]
@vidstige ,
errors on test case
because, the GlyphPathBuilder
dose not use the  contour, line-segment etc.

err3_on_net20

what should I do?

File: NOpenType/GlyphPathBuilderBase.cs (hunk 14 above)
Comment (prepare) [reply NOpenType/GlyphPathBuilderBase.cs:110]: ok

File: NOpenType/GlyphPathBuilderBase.cs (hunk 15 above)
Comment (prepare) [reply NOpenType/GlyphPathBuilderBase.cs:128]: ok

**Comment by vidstige** [comment]
Alright, your too slow I pushed some fixes. I'll fix the merge problem now,

**Comment by vidstige** [comment]
Alright, it all seems to work now. Compiles and unit tests green! ✅  I also rebased for cleaner history in #16. Enjoy! :-)

**Comment by vidstige** [comment]
Sorry if I seem harsh in my messages. It's just my baby project and I want to keep it clean and slick. But I do appreciate all your contributions. Hope you understand! :-)
